File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        // split s into a list of words
        String[] words = s.split(" ");

        // check if s follows the pattern
        if (pattern.length() != words.length) {
            return false;
        }
        Map<Character, String> charToWord = new HashMap<>();
        Map<String, Character> wordToChar = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String w = words[i];
            if (charToWord.containsKey(c) && !charToWord.get(c).equals(w)) {
                return false;
            }
            if (wordToChar.containsKey(w) && !wordToChar.get(w).equals(c)) {
                return false;
            }
            charToWord.put(c, w);
            wordToChar.put(w, c);
        }
        return true;
    }
}

File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        stringstream ss(s);
        vector<string> words;
        string word;
        while (ss >> word) {
            words.push_back(word);
        }

        // check if s follows the pattern
        if (pattern.size() != words.size()) {
            return false;
        }
        unordered_map<char, string> char_to_word;
        unordered_map<string, char> word_to_char;
        for (int i = 0; i < pattern.size(); i++) {
            char c = pattern[i];
            string w = words[i];
            if (char_to_word.count(c) && char_to_word[c] != w) {
                return false;
            }
            if (word_to_char.count(w) && word_to_char[w] != c) {
                return false;
            }
            char_to_word[c] = w;
            word_to_char[w] = c;
        }
        return true;
    }
};